_posts/2023-06-04-not-generating-constructors.md

```diff
@@ -0,0 +1,411 @@
+    fmt::println("{}", s[0]);
+}
+
+void take_span_ref(const std::span<const int>& s) {
+    fmt::println("{}", s[0]);
+}
+
+void take_vec(const sus::Vec<int>& v) {
+    fmt::println("{}", v[0]);
+}
+
+void take_slice(sus::Slice<int> s) {
+    fmt::println("{}", s[0]);
+}
+
+void take_slice_ref(const sus::Slice<int>& s) {
+    fmt::println("{}", s[0]);
+}
+
+int main() {
+    std::vector<int> a = {1, 2, 3, 4, 5, 6, 7};
+    take_vector(a);
+    take_pointer(a.data(), a.size());
+    take_span(a);
+    take_span_ref(a);
+
+    auto b = sus::Vec<int>::with_values(1, 2, 3, 4, 5, 6, 7);
+    take_vec(b);
+    take_slice(b);
+    take_slice_ref(b);
+}
+```
+
+We've added a call to `take_span_ref()` which receives a `const std::span&`. Unsurprisingly this
+does not improve anything, here is the codegen for that call:
+```asm
+lea    rdx,[rbp-0x60]
+lea    rax,[rbp-0x30]
+mov    rsi,rdx
+mov    rdi,rax
+call   404b2e <std::span<int const, 18446744073709551615ul>::span<std::vector<int, std::allocator<int> >&>(std::vector<int, std::allocator<int> >&)>
+lea    rax,[rbp-0x30]
+mov    rdi,rax
+call   403acf <take_span_ref(std::span<int const, 18446744073709551615ul> const&)>
+```
+We have to construct a `std::span` from the `std::vector` in order to give a reference to it in the
+function call, so this comes out at 8 instructions, without the `std::span` constructor being
+inlined.
+
+Now let's look at the Subspace types.
+
+First, receiving a `const sus::Vec&` looks the same as `const std::vector&`, which isn't surprusing:
+```asm
+lea    rax,[rbp-0x70]
+mov    rdi,rax
+call   403c75 <take_vec(sus::Vec<int> const&)>
+```
+We grab the pointer, store it and call the function, for three instructions.
+
+And receiving a `sus::Slice` looks much like receiving a `std::span`:
+```asm
+lea    rax,[rbp-0x70]
+mov    rdi,rax
+call   404bf0 <sus::Vec<int>::operator sus::Slice<int>&() &>
+mov    rdx,QWORD PTR [rax]
+mov    rax,QWORD PTR [rax+0x8]
+mov    rdi,rdx
+mov    rsi,rax
+call   403e1b <take_slice(sus::Slice<int>)>
+```
+We construct a `sus::Slice` and then pass that by value to `take_slice()`. This actually comes in
+two instructions shorter than the call to `take_span()`, but it's still way more codegen than
+passing the `const sus::Vec&` directly.
+
+But lastly, we have the call to `take_slice_ref()`, which now differs significantly from the
+equivalent `take_span_ref()` with the standard library types:
+```asm
+lea    rax,[rbp-0x70]
+mov    rdi,rax
+call   404bf0 <sus::Vec<int>::operator sus::Slice<int>&() &>
+mov    rdi,rax
+call   403fd1 <take_slice_ref(sus::Slice<int> const&)>
+```
+Now we're down to 5 instructions. We call the conversion operator to `sus::Slice&`, which returns
+_a pointer_ that we store, and then call `take_slice_ref()` with it.
+
+But why is this operator returning _a pointer_, instead of a `sus::Slice` by value? The trick here
+is that `sus::Vec` is implemented by holding a `sus::SliceMut` inside it, which holds a

```

Comment:
This is a tangent for this article, but it might merit a link to a design doc, or a footnote: I would have expected that `SliceMut` would be the most primitive type of slice, and that the immutable `Slice` would contain a `SliceMut` whose mutation member functions it simply doesn't expose.

Reply:
It's really just because of the conversion direction. A SliceMut can be used as a Slice but not vice versa. So SliceMut must contain the Slice to return a reference to it which is the conversion operation.

Reply:
Footnoted. We'll see if my blog can render that okay.